### ISSUE TITLE
Installation errors on the command line

### ISSUE THREAD

**Issue by thundernixon** [googlefonts/tools#40]
I'm trying to work with a Google Font and I'm running into issues with setting up gftools to do so.

Specifically, `sudo easy_install pip` went fine, but then `pip install --upgrade gftools` ran partially, but failed with the following error:

```
OSError: [Errno 1] Operation not permitted: '/System/Library/Frameworks/Python.framework/Versions/2.7/share'
```

I tried to use `sudo chown stephennixon /Library/Python/2.7/site-packages` to get past this on the suggestion of a Stack Overflow post, then ran `pip install --upgrade gftools` again. 

Now, I'm getting this error: 

```
Error: [('/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/markers.pyc', '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/markers.pyc', "[Errno 1] Operation not permitted: '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/markers.pyc'"), ('/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/__init__.py', '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/__init__.py', "[Errno 1] Operation not permitted: '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/__init__.py'"), ('/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/markers.py', '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/markers.py', "[Errno 1] Operation not permitted: '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/markers.py'"), ('/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/__init__.pyc', '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/__init__.pyc', "[Errno 1] Operation not permitted: '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib/__init__.pyc'"), ('/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib', '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib', "[Errno 1] Operation not permitted: '/var/folders/tq/g3lct8f965z_nw5y9fxbg7_h0000gn/T/pip-5aajiB-uninstall/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/_markerlib'")]
```

I've Googled / Stack Overflowed around for the solution, plus looked around the issues on this repo, but nothing's standing out. I'm new to Python, so I'm a bit unsure of how to proceed. Any suggestions would be greatly appreciated!

**Comment by anthrotype** [comment]
TL;DR: never use `sudo pip install ...` with the system python. 

you are attempting to install a third-party module using the pip installer to the Apple system python. The latter default installation directory is inside a location that is locked by the so-called SIP protection. Basically Apple (rightly so) doesn't like users to fiddle with the system folders.

I recommend you install another python distribution that is kept separate from the operating system's.

The simplest is to download the latest python from https://www.python.org/downloads/
I believe this gftools only support python 2.7 for now.
pip comes  pre-installed with this, and you'll not longer need to *sudo* pip install because the folder where this installs third party packages will be writable by regular user and will not conflict with the system's python.

Another way is to install [homebrew](https://brew.sh/), the package manager for macOS, and then `brew install python` (or `brew install python3`). Same here, you'll get pip (or pip3) and you won't need sudo.

**Comment by anthrotype** [comment]
there's also the option to use `pip install --user ...`, but it looks like you somehow broke your system python installation -- this is what happens when one copy pastes snippets from Stack Overfllow, especially those that start with "sudo..." :-P

much better to leave /usr/bin/python alone is his misery.

**Comment by anthrotype** [comment]
sorry, I didn't write the readme :-)

I think the readme should be updated, but I think it's not the responsibility of this or any other python module to explain users how to set up their python environment. They should be redirected to https://docs.python.org or https://packaging.python.org/guides/installing-using-pip-and-virtualenv/

No, of course, you can't uninstall the system python.. If anything, you can wipe and reinstall the OS!

Try to uninstall the homebrew python and re-install it again. Then do `which python` and ensure it reads `/usr/local/bin/python` and not `/usr/bin/python` (the latter is the system python).

otherwise check your $PATH variable to make sure the right python binary is run when you type `python` in the shell.

**Comment by anthrotype** [comment]
ah, another thing.

type `pip --version` and ensure that that `pip` executable is actually linked to the homebrew python, and _not_ to the system python.

if it's wrong, then re-installa pip from the homebrew python so that the old pip script in /usr/local/bin is replaced by the new one correctly linked with the homebrew python.

```sh
$ /usr/local/bin/python -m pip install --ignore-installed pip
```
if it fails with permission errors, use sudo the last time. That;s because you previously installed it with `sudo easy_install pip`

**Comment by anthrotype** [comment]
basically the pip script by default installs to the site-packages folder of the python interpreter that was used to install the pip installer. 

that's why sometimes you see people recommending to do `python -m pip` instead of just `pip`, because with the former incantation you are certain that you're calling the pip module of that python exectutable, and not a generic pip which may not match the python that you expected (if it has been overwritten at some point)

**Comment by felipesanches** [comment]
Given @thundernixon's trouble, which install instructions should we provide on readme to make this outcome less likely for future users, @anthrotype ?

**Comment by davelab6** [comment]
Felipe please make a PR to update the README :)

**Comment by chrissimpkins** [comment]
@felipesanches pip should be default in current Python releases (including system installs of Python2).  Should not need a separate pip install any longer.

**Comment by felipesanches** [comment]
I use virtualenvs during development. But I think it would likely complicate matters for people who may be not so happy about command line usage, as it would require an extra step of activating the venv before first use in each terminal session

**Comment by anthrotype** [comment]
BTW, easy_install is now deprecated as an installation tool (replaced by pip itself).
Nowadays, pip comes pre-installed with most python distributions, such as the official python.org or homebrew.
To bootstrap pip, one should either do `python -m ensurepip` or use the [get-pip.py](https://bootstrap.pypa.io/get-pip.py) script

**Comment by anthrotype** [comment]
it depends on your user base. if they are developers, they should know how to (find out how to) make sure pip is present and is linked to the right python executable, etc.
A link to the official documentation should do.

on the other hand, if you wish to make it more widely available.. maybe you could provide self-contained executables (e.g. with pyinstaller) that don't require python to be there.

It's hard to support all possible platforms and use cases, that's why there is the official python.org and PyPA documentation.

**Comment by felipesanches** [comment]
Does it mean the ideal install instructions in readme would be something like this?

```bash
python -m ensurepip
pip install --upgrade gftools
```

The intended user base is likely type designers who are barely comfortable with using the command line (after all gftools is a set of cmd line tools), but may not be experts at it, since their core skill is type-design instead.

**Comment by anthrotype** [comment]
> including system installs of Python2

@chrissimpkins you mean Apple python?! I would be surprised if that was the case...

**Comment by chrissimpkins** [comment]
> I use virtualenvs during development. But I think it would likely complicate matters for people who may be not so happy about command line usage, as it would require an extra step of activating the venv before first use in each terminal session

It doesn't appear that you are pinning versions on any package dependencies here, but your virtualenv suggestion here would be a "best practices" approach...  I think that most who write install documentation for Python modules assume that users understand/recognize the venv issue.

**Comment by felipesanches** [comment]
I would guess that type designers most likely lack the knowledge of setting up and using virtualenvs, though...

**Comment by anthrotype** [comment]
maybe `python -m ensurepip --user` so you don't need sudo... but then the pip script will go to the users' bin folder which is not in the $PATH..., so you either tell people to call it like `python -m pip` or you go again telling people how to set up their ~/.bashrc and what not...

I'm sorry life is complicated.

**Comment by anthrotype** [comment]
I know some linux distros (don't know about Apple) chop off the ensurepip module. You have to understand that pip may be seen as a threat by operating system vendors because it potentially lead users to unknowingly install malicious software

good news is pip 10 will default to `pip instal --user` all the time

**Comment by felipesanches** [comment]
Should we then consider a shell script for setting up a user installation with a single command ? I'm trying to hold the hands of the users here...

**Comment by felipesanches** [comment]
I've been to a friend's house last week to debug the install instructions of Fontbakery 0.3.3 on a MacOS. I can guarantee you that he was not very comfortable with all the typing that was going on. I think it all scares lots of people. :-P

**Comment by anthrotype** [comment]
you have to test on a real vanilla mac system what are the steps required to go from nothing to a working python setup.

I think recommending users to use homebrew python is a sensible solution here. You get an up-to-date python, pip is there by default, and they will be both in the PATH, and users can pip install without sudo.

**Comment by felipesanches** [comment]
At some point he asked me if we couldn't simply provide a ".DMG" :-P I understood the damend but I was unable to properly reply to his question. (and it sounded a bit weird to me, since both fontbakery and gftools are command line tools after all...)

**Comment by anthrotype** [comment]
> I think it all scares lots of people. :-P

then maybe the self-contained executables with pyinstaler would be good for them.
they just go to github, download the zip from the latest release page, unzip it and run the script. No python needed

**Comment by chrissimpkins** [comment]
> @chrissimpkins you mean Apple python?! I would be surprised if that was the case...

@anthrotype Just checked and you are correct.  I thought that this came with the Apple Python distribution once they began pushing versions 2.7.9+ in OS X updates...

**Comment by anthrotype** [comment]
yeah.. They don't care about :snake:  anymore now that they have :hatched_chick:

**Comment by chrissimpkins** [comment]
@felipesanches this works very well to turn scripts into simple GUI applications on OSX Felipe

http://sveinbjorn.org/platypus/ 

An example with a dmg installer (and the scripts that show how I made the dmg installer) are included in this repo:

https://github.com/source-foundry/Woffle

It doesn't work well extensive user interaction but you can use it for simple DnD tasks if that would work here.

**Comment by anthrotype** [comment]
@chrissimpkins that's clever, though it would not solve the python bootstrapping issue. Platypus just makes an .app wrapper for a shell script. Unless you want to embed a full python interpreter with the full stack of third-party modules (which is what pyinstaller would do automatically).

the drag-and-drop thing is a bit beyond my personal taste.

**Comment by chrissimpkins** [comment]
> full stack of third-party modules 

Would be better suited to individual applications for each task (i.e. 1:1 Python script : GUI application), but agree with you.  Agree with you about the DnD GUI comment as well, however, my guess is that there is a user base out there that appreciates this simplicity...

**Comment by chrissimpkins** [comment]
Maybe Google would provide some servers for a set of Python web applications? :)

**Comment by felipesanches** [comment]
I think that the reality is that gftools is primarily used by the Google Fonts team which certainly knows how to setup and install python programs on the command line. But then other users end up getting interested (given the usefulness of the tools) and using them as well.

So, we might as well want to give a helping hand to these users.

**Comment by felipesanches** [comment]
On the other hand, I am not sure how much effort would be justifiable here, in addition to minimal working install instructions on README.

**Comment by felipesanches** [comment]
Having said that, Platypus sounds cool, but I think that's too much. I'm not willing to follow that path.

**Comment by thundernixon** [comment]
Just to chime in on the conversation as one user, I do like the command line for a few things: I use it for git, and I've used it for running some web-dev build processes in the past. For this, I do have some level of experience with Homebrew and npm. However, I haven't yet done a lot of other types of software development yet, and my current type design knowledge is mostly around drawing with beziers, rather than Python production techniques. I'm working on this, but of the type designers I know, only the nerdier ones really dive into a lot of development. :)

**Comment by felipesanches** [comment]
I'm not sure yet what's the best thing we can write on our README here.

**Comment by thundernixon** [comment]
Well, after chasing this around, I _think_ I got gftools to install without error, though `which python` still points to `/usr/bin/python` and not `/usr/local/bin/python`. I'll try the builds I'm hoping to do, and see whether they work. Thanks for your help and input!

**Comment by felipesanches** [comment]
Thanks for reporting the issue here :-)

**Comment by chrissimpkins** [comment]
> Just to chime in on the conversation as one user, I do like the command line for a few things: I use it for git, and I've used it for running some web-dev build processes in the past. For this, I do have some level of experience with Homebrew and npm. However, I haven't yet done a lot of other types of software development yet, and my current type design knowledge is mostly around drawing with beziers, rather than Python production techniques. I'm working on this, but of the type designers I know, only the nerdier ones really dive into a lot of development. :)

@thundernixon that's an extremely important point and suggests that much of the excellent command line based open source tooling for type development may not be accessible to a group of potential users who could benefit from these projects.  Out of interest, do you and your design colleagues have a desire to become facile on the command line (in which case some an education approach would work best so that you can use current tool formats) or would you prefer to have tools that feel more familiar (based upon your comments, it sounds as though this means GUI style tools where you can interact with a graphical interface)?

**Comment by chrissimpkins** [comment]
> I'm not sure yet what's the best thing we can write on our README here.

@felipesanches call me old school :) , but would download (including submodule) + `python setup.py install` as a second recommended installation approach work for users who don't have (or don't want to install) pip?

**Comment by chrissimpkins** [comment]
> Well, after chasing this around, I think I got gftools to install without error, though which python still points to /usr/bin/python and not /usr/local/bin/python. 

Seems like it may be time to report this on the Homebrew repository if it hasn't been.  This is happening to many of us on OS X and `brew link` is not fixing it.

![python](https://user-images.githubusercontent.com/4249591/33347881-6105bb40-d462-11e7-8160-84a587cb7728.png)

**Comment by anthrotype** [comment]
python setup.py install will use the deprecated easy_install.py module from setuptools, will create an .egg instead of a wheel package, will attempt to resolve dependencies via easy_install and thus build .eggs from the source distributions (zip or tar.gz) available on PyPI, but the latter may fail in case of packages that require a C/C++ compiler because they include extension modules, and the compiler is missing on the user's environment.
That's why I always recommend to install with pip, even from a git source tree (`pip install -e .`). 

in short, pip + wheels are the best way to install third party modules.

**Comment by anthrotype** [comment]
if you `echo $PATH`, is the /usr/local/bin folder there, and is it _before_ /usr/bin?

**Comment by chrissimpkins** [comment]
Argues that a package manager should be part of every distribution of Python. At the very least, when you install developer command line tools with XCode...

**Comment by chrissimpkins** [comment]
> if you echo $PATH, is the /usr/local/bin folder there, and is it before /usr/bin?

yep...

**Comment by chrissimpkins** [comment]
Problem was that the symlink in `/usr/local/bin` disappeared at some stage of updates (likely OS based upon the timeframe when this occurred).  Here is the workaround that I used to manually create the python symlink in `/usr/local/bin` for anyone else having the same issue:

```
$ ln -s ../Cellar/python/2.7.14/bin/python2 /usr/local/bin/python
```

Now `python` reports proper Homebrew version of Python2 (2.7.14).

**Comment by anthrotype** [comment]
it looks like it was intentional
https://github.com/Homebrew/homebrew-core/commit/e7f4e1c03abbe0842b77e3d163e737a81ca955e7

homebrew python no longer shadows the system python...

**Comment by anthrotype** [comment]
```
 This formula installs a python2 executable to #{HOMEBREW_PREFIX}/bin.
 If you wish to have this formula's python executable in your PATH then add
 the following to #{shell_profile}:
    export PATH="#{opt_libexec}/bin:$PATH"
```

**Comment by chrissimpkins** [comment]
Ha! Guess it makes sense to RTFD every once in awhile...  You would think that they would have provided some warnings with python updates before moving in that direction...

**Comment by anthrotype** [comment]
I just reinstalled it and I got this 

```
==> Caveats
This formula installs a python2 executable to /usr/local/bin.
If you wish to have this formula's python executable in your PATH then add
the following to ~/.bash_profile:
  export PATH="/usr/local/opt/python/libexec/bin:$PATH"

Pip and setuptools have been installed. To update them
  pip2 install --upgrade pip setuptools

You can install Python packages with
  pip2 install <package>

They will install into the site-package directory
  /usr/local/lib/python2.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python.html
```

**Comment by chrissimpkins** [comment]
Ah, OK I blame myself then.  It may have scrolled across the screen during a recent Python version update and I didn't catch it.  Thanks for locating that Cosimo.  That is the permanent fix.

**Comment by anthrotype** [comment]
yeah, though telling people to add export PATH=... to ~/.bash_profile is not very type-designer-friendly based on our previous discussion. Anyway, we spent way too much time on this.
Just point people to the official python and pip docs, and they'll figure out by themselves like we all did (ie. google, stack overflow and a lot of patience)

**Comment by felipesanches** [comment]
Yeah... this thread ended up becoming a chat room LOL

Let's try to refocus on answering this objective question:

what do we write on the README of this repo to make it safe and harmless for (potentially clueless) users to install gftools without messing up their systems?

**Comment by chrissimpkins** [comment]
Sorry... :)

**Comment by anthrotype** [comment]
> what do we write on the README

remove all instructions, link to official docs, not your job to tell ppl how to install/setup their python

**Comment by felipesanches** [comment]
yeah.... maybe.... I guess some hand holding may still be necessary though.

**Comment by felipesanches** [comment]
As the current instructions are a bit misleading I could start by simply removing that.

**Comment by thundernixon** [comment]
I'm always a fan of providing a best practice install route, if it's possible. Maybe you could say something like...

===

If you already have pip installed, simply run `pip install --upgrade gftools`. 

If you still need to set up Python on your usr/bin/local and you have homebrew, you can run `brew install python` to set that up, then use the first command to install gftools. 

If you run into issues with installation, you may need to edit your `~/.bash_profile` to include `export PATH="/usr/local/opt/python/libexec/bin:$PATH"`. If you are still having trouble, you may need to refer to the official python and pip docs for further instructions on setting up your environment.

===

👆 Sorry, still not 100% if that's accurate information.

That said, I could understand if you just link to the official docs and don't give recommendations here. Ultimately it's a decision around who you see as your target audience and how accessible you want this tool to be for beginners. I personally am happy to work within the command line and don't always require GUIs, but it can be difficult to know how to get started on a new toolchain. I could read through the docs, but it might still be difficult to understand how that information connects to gftools, and the main thing I really want to do in my limited available time is to test out working with / expanding with a typeface from the library.

**Comment by thundernixon** [comment]
All in all, things do seem to be working now, though I have that lingering feeling that I did something hacky, and I don't know how critical it is that `which python` is still returning `/usr/bin/python`. 

I know that I just need to spend some more time getting to know the ins and outs of working with Python, and I really appreciate all the discussion and information here. I'm sure that what doesn't make sense immediately will come in handy at some point in the future. :)

**Comment by davelab6** [comment]
If `which python` still isn't giving you homebrew'd python, edit your bash profile as your suggested docs suggest :D

Your suggestion LGTM

**Comment by anthrotype** [comment]
maybe recommending the official python.org distribution (which comes with a familiar dpkg installer, and as far as I remember does shadow the system python and installs itself in the PATH) would be better for non-experienced users.
if one knows about homebrew, they are also more likely to know how to know about command line tools, PATH, .bashrc, etc.
homebrew python is just one among many third party python distributions. It just happens to be what many of us use, but it's certainly not the only way to run python on mac.

**Comment by thundernixon** [comment]
> If which python still isn't giving you homebrew'd python, edit your bash profile as your suggested 

I'm trying to figure out why it just worked when it wasn't before ... I guess the "Caveats" post above had a better suggested line for the bash profile that other blog posts I had found.

True, if the package download from python.org includes Pip and would enable the gftools install to go properly, that seems like definitely the simplest option for those coming in without an opinion on how to set up Python. And those who do have an opinion are (myself excluded 😄) more able to use and troubleshoot homebrew or their method of choice.

**Comment by anthrotype** [comment]
yes, just tried using the official installer and by default it does the right thing ™️ 
don't even need to go into "Customize" panel. Just click continue continue continue... 💥  done

**Comment by chrissimpkins** [comment]
> All in all, things do seem to be working now, though I have that lingering feeling that I did something hacky, and I don't know how critical it is that which python is still returning /usr/bin/python.

@thundernixon if you attempt to execute anything that you install with pip using the `python` executable it will fail b/c it will not be able to import the pip installed modules.  They will be linked to your Homebrew version of Python which the Homebrew developers have chosen to symlink to `python2` (and not `python`) as of recent releases.  This is new behavior.  It used to symlink around the system version of Py2 and execute the Homebrew version of Py2 interpreter when you used `python` on the command line.  If you choose not to add the symbolic link to the Homebrew Python version or modify your $PATH to default to the Homebrew installed version as per Cosimo's post above, you can directly execute python scripts with `python2 [python script]` instead of `python [python script]` and it will pick up any packages that you have installed with pip.  I've run into problems with imports in executables with that approach and it isn't recommended unless you absolutely want to avoid symlinking `python` with the Homebrew version.

Alternatively, install python3 with `brew install python3` and execute everything with the Py3 interpreter instead of the Py2 interpreter.  There will be no conflicts with system Python installs there (Py3 is not currently installed by default on OS X) and executable packages *should* pick up all imports without issues.  Python3 uses `pip3` as the package manager instead of `pip`.

It would look something like this:

```
$ brew install python3
$ which python3
/usr/local/bin/python3
$ python3 --version
Python 3.6.3
$ pip3 install --upgrade [PyPI package]
```
